pkg/controller/traceflow/validate.go

```diff
@@ -0,0 +1,111 @@
+		resp := &admv1.AdmissionResponse{
+			UID:     review.Request.UID,
+			Allowed: allowed,
+		}
+		if !allowed {
+			resp.Result = &metav1.Status{
+				Message: deniedReason,
+			}
+		}
+		return resp
+	}
+
+	klog.V(2).InfoS("Validating Traceflow", "request", review.Request)
+
+	var newObj, oldObj crdv1alpha1.Traceflow
+	if review.Request.Object.Raw != nil {
+		if err := json.Unmarshal(review.Request.Object.Raw, &newObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing current Traceflow")
+			return newResponse(false, err.Error())
+		}
+	}
+	if review.Request.OldObject.Raw != nil {
+		if err := json.Unmarshal(review.Request.OldObject.Raw, &oldObj); err != nil {
+			klog.ErrorS(err, "Error de-serializing old Traceflow")
+			return newResponse(false, err.Error())
+		}
+	}
+
+	switch review.Request.Operation {
+	case admv1.Create:
+		klog.V(2).InfoS("Validating CREATE request for Traceflow")
+		allowed, deniedReason := c.validateCreate(&newObj)
+		return newResponse(allowed, deniedReason)
+	case admv1.Update:
+		klog.V(2).InfoS("Validating UPDATE request for Traceflow")
+		allowed, deniedReason := c.validateUpdate(&oldObj, &newObj)
+		return newResponse(allowed, deniedReason)
+	case admv1.Delete:
+		// This shouldn't happen with the webhook configuration we include in the Antrea YAML manifests.
+		klog.InfoS("Validating DELETE request for Traceflow. This shouldn't happen with the webhook configuration we include in the Antrea YAML manifests.")
+		return newResponse(true, "")
+	default:
+		klog.InfoS("Unknown request type for Traceflow")
+		return newResponse(false, "unknown request type for Traceflow")
+	}
+}
+
+func (c *Controller) validateCreate(tf *crdv1alpha1.Traceflow) (allowed bool, deniedReason string) {
+	if tf.Spec.Destination.Service != "" && !features.DefaultFeatureGate.Enabled(features.AntreaProxy) {
+		return false, "using Service destination requires AntreaProxy feature enabled"
+	}
+	if tf.Spec.Destination.IP != "" {
+		destIP := net.ParseIP(tf.Spec.Destination.IP)
+		if destIP == nil {
+			return false, fmt.Sprintf("destination IP is not valid: %s", tf.Spec.Destination.IP)
+		}
+	}
+	if !tf.Spec.LiveTraffic {
+		srcPod, err := c.podLister.Pods(tf.Spec.Source.Namespace).Get(tf.Spec.Source.Pod)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				err = fmt.Errorf("requested source Pod %s not found", k8s.NamespacedName(tf.Spec.Source.Namespace, tf.Spec.Source.Pod))
+			}
+			return false, err.Error()
+		}
+		if srcPod.Spec.HostNetwork {
+			return false, "using hostNetwork Pod as source in non-live-traffic Traceflow is not supported"
+		}
+	}
+	if tf.Spec.Source.Pod == "" && tf.Spec.Destination.Pod == "" {
+		return false, fmt.Sprintf("Traceflow %s has neither source nor destination Pod specified", tf.Name)
+	}
+	return true, ""
+}
+
+func (c *Controller) validateUpdate(oldTf, newTf *crdv1alpha1.Traceflow) (allowed bool, deniedReason string) {
+	return c.validateCreate(newTf)
+}
```

Comment:
for better readability and simplicity, you can rename `validateCreate` to `validate` or `validateTraceflow`, and call it in `Validate` directly. There is no point to decode `oldObj` since it's never used.

Reply:
My consideration is that the validation logic for updates and creations is indeed different because when validating an update, we may need to consider whether the spec before and after the modification are consistent. Or I think maybe we can prohibit updating traceflow, as it seems to make no sense.

Reply:
If they could be different, it makes less sense to call validateCreate in validateUpdate? 

You can always change the code and decode the old object when implementing more validation. But if you just have that in mind and never implement it, some useless code will be kept there perhaps forever and may confuse other contributors, can we keep only necessary code in this commit?

Reply:
Indeed it makes no sense to call validateCreate in validateUpdate, so I think the original logic has problems, but in this PR, I prefer to keep the logic unchanged.

Of course we can keep the lines of code as small as possible, but I think a little bit of "over-design" won't cause confusion.